### PR TITLE
detector class major rework

### LINKED ISSUE #259: Conversion of DCYL detector gives broken files
I've generated some energy-depth spectra for 200 MeV proton beams using DCYL scorer and SH12A v0.7.1.

Using:
```
*----0---><----1---><----2---><----3---><----4---><----5---><----6--->
DCYL             0.0       0.0       0.0      10.0       7.0      30.0
                   1         1       300        -1      DOSE    ex_dcyl_dose_en_p
                 0.0     220.0       300              ENERGY
```

I obtain:

ex_dcyl_dose_en_p

Clearly is seen that the depth and energy axis are swapped.

If I change number of bins in energy and depth then it gets completely broken:
```
*----0---><----1---><----2---><----3---><----4---><----5---><----6--->
DCYL             0.0       0.0       0.0      10.0       7.0      30.0
                   1         1       300        -1      DOSE    ex_dcyl_dose_en_p
                 0.0     220.0       220              ENERGY
```

ex_dcyl_dose_en_p

### PR DESCRIPTION
**Pull request by grzanka**
Fixes #259 , #237 , #232 

This PR has started as bugfix of #259 , but in the course of fixing it, I decided to rework completely a detector class structure.

The aim was to provide some simple API for readers and writer class and simplify processing of input files/directories

### CODE REVIEW

**Comment by grzanka** [comment]
Now looks much better, with:
```
*----0---><----1---><----2---><----3---><----4---><----5---><----6--->
DCYL             0.0       0.0       0.0      10.0       7.0      30.0
                   1         1       300        -1      DOSE    ex_dcyl_dose_en_p
                 0.0     220.0       300              ENERGY
```
we get:
ex_dcyl_dose_en_p


With:
```
*----0---><----1---><----2---><----3---><----4---><----5---><----6--->
DCYL             0.0       0.0       0.0      10.0       7.0      30.0
                   1         1       300        -1      DOSE    ex_dcyl_dose_en_p
                 0.0     220.0       220              ENERGY
```

we get now:
ex_dcyl_dose_en_p

**Comment by nbassler** [comment]
also the colourbar: I would expect rather some histogramming kind of unit here [dN/dE]. (this is what DCYL should return, I think)

**Comment by nbassler** [comment]
or dN/dD [(MeV/g/primary)^-1]

**Comment by grzanka** [comment]
OK, I'll fix the axis labels and try to swap the axis

**Comment by grzanka** [comment]
@nbassler speaking about units - they are like histogramming, but the unit depends on the MSH or CYL scorer.
What do you think about simply writing `dE [(MeV/g/primary) per bin]` ?
Then the reader should guess the bin type and units by looking at axis description, here the bin units will be `MeV x cm`.

For other scorers the bin can be i.e. `MeV x mrad`

**Comment by nbassler** [comment]
not sure what you meant by "dE [(MeV/g/primary) per bin]"?
Histograms are differential with respect to the x-axis. dN/dx vs x.

**Comment by grzanka** [comment]
We have two types of histograms in differential scoring from which we can generate a plot:
1-D (drawn with lines) where we have dN/dx vs x and 2-D (drawn with heatmap) where we have d^2N/ dx dy (color scale) vs x and y  or dN /dS (where S = x*y)

**Comment by nbassler** [comment]
The plots you showed earlier are 1D histograms at a range of positions. This is not the same as double differential historgrams.

**Comment by grzanka** [comment]
Yes, I see the point - in one dimension the unit is per MeV (there was a division by bin width applied) and in the other per bin (no division per bin width applied) - is that correct ?

**Comment by nbassler** [comment]
yes, I think so.

File: pymchelper/detector.py [hunk 1]
```diff
@@ -1,37 +1,37 @@
...
-from pymchelper.shieldhit.detector.detector_type import SHDetType
-from pymchelper.shieldhit.detector.estimator_type import SHGeoType
-from pymchelper.writers.excel import ExcelWriter
-from pymchelper.writers.inspector import Inspector
-from pymchelper.writers.plots import ImageWriter, GnuplotDataWriter, PlotDataWriter
-from pymchelper.writers.shieldhit import TxtWriter
-from pymchelper.writers.sparse import SparseWriter
-from pymchelper.writers.trip98 import TripCubeWriter, TripDddWriter
 
 logger = logging.getLogger(__name__)
```
Comment (ant6) [inline pymchelper/detector.py]: I think you never use logger in this file 😉

File: pymchelper/io.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        # not needed if user requested not to include errors
+        if error != ErrorEstimate.none:
+            __M2 = np.zeros_like(result.data_raw)
+
+        # loop over second and next files, if present
+        for counter, filename in enumerate(input_file_list[1:]):
+            next_det = fromfile(filename)
+
+            # Running variance algorithm based on algorithm by B. P. Welford,
+            # presented in Donald Knuth's Art of Computer Programming, Vol 2, page 232, 3rd edition.
```
Comment (ant6) [inline pymchelper/io.py]: 👍

File: pymchelper/io.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import numpy as np
+
+from pymchelper.detector import Detector, average_with_nan, ErrorEstimate
+from pymchelper.readers.common import guess_reader, group_input_files
+from pymchelper.writers.common import Converters
+
+logger = logging.getLogger(__name__)
+
+
+def fromfile(filename):
```
[the remark this answers is not in the document]
Comment (grzanka) [reply pymchelper/io.py]: The functions with `_` are hidden/private, this one here is public and part of API. In the naming I was trying to mimic numpy, see:
 - https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.fromfile.html
 - https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.loadtxt.html#numpy.loadtxt

Nevertheless I'll add some comments explaining what this method does.

File: pymchelper/readers/shieldhit.py [hunk 4]
```diff
@@ -4,72 +4,106 @@
...
+        name = ("Radius (R)", "Angle (PHI)", "Position (Z)")[axis]
+    else:
+        name = ""
 
     # dirty hack to change the units for differential scorers
-    if hasattr(detector, 'dif_axis') and hasattr(detector, 'dif_type'):
+    if hasattr(detector, 'dif_axis') and hasattr(detector, 'dif_type') and axis == detector.dif_axis:
         if detector.dif_type == 1:
-            diff_unit = SHBinaryReader.get_detector_unit(SHDetType.energy, detector.geotyp)[0]
+            unit, name = _get_detector_unit(SHDetType.energy, detector.geotyp)
```
[the remark this answers is not in the document]
Comment (grzanka) [reply pymchelper/readers/shieldhit.py]: Thanks - I'll fix it !

File: pymchelper/detector.py [hunk 5]
```diff
@@ -40,380 +40,93 @@ class ErrorEstimate(IntEnum):
...
-        return self.axis_values(Axis.x)
-
-    @property
-    def y(self):
-        return self.axis_values(Axis.y)
-
-    @property
-    def z(self):
-        return self.axis_values(Axis.z)
+        plotting_order = (0, 1, 2)
```
[the remark this answers is not in the document]
Comment (grzanka) [reply pymchelper/detector.py]: I think it's a mess generated by github. The changeset was so large, that the diffs are not displayed correctly. See full file here:
https://github.com/DataMedSci/pymchelper/blob/feature/259-dcyl/pymchelper/detector.py

I'll also explain the purpose of the code in the comment.

File: pymchelper/detector.py (hunk 1 above)
Comment (grzanka) [reply pymchelper/detector.py]: But I will

**Comment by nbassler** [comment]
Might close #205 as well.
Wonder if DPLANE will work.

**Comment by grzanka** [comment]
I will check DPLANE, single differential scoring works for LET and ENERGY.

**Comment by nbassler** [comment]
and may also close #203  ? :)

**Comment by grzanka** [comment]
yes, it may also close #203.
except for one case when we have 3D mesh + one differential variable - Detector class is limited to hold only 3D data.

**Comment by grzanka** [comment]
@nbassler ready for merge once the tests are done

**Pull request merged**